File: api-test/perf/concurrent_heavy.py
import os
import time
import asyncio
import httpx
import random
from common import (
    get_unique_data,
    MGMT_URL,
    CACHE,
    BASE_URL,
    KEY,
)

# --- Configuration ---
CONCURRENCY = 20
FILE_SIZE_MB = 100

# --- Async Utilities (for Concurrent Test) ---


async def async_init_upload(client):
    url = f"{MGMT_URL}/api/v1/cache/{CACHE}/multipart-nar?compression=xz"
    res = await client.post(url)
    res.raise_for_status()
    return res.json()["uploadId"]


async def async_perform_upload(client, upload_id, data, timeout=120):
    url = f"{BASE_URL}/{upload_id}"
    res = await client.put(url, content=data, timeout=timeout)
    res.raise_for_status()
    return res


async def async_complete_upload(
    client, upload_id, file_hash, size, suffix, timeout=120
):
    store_hash = file_hash[:32]
    url = f"{MGMT_URL}/api/v1/cache/{CACHE}/multipart-nar/{upload_id}/complete"
    body = {
        "narInfoCreate": {
            "cFileHash": f"{file_hash}",
            "cFileSize": size,
            "cStoreHash": store_hash,
            "cStoreSuffix": f"perf-test-{suffix}",
            "cNarHash": f"sha256:{file_hash}",
            "cNarSize": size,
            "cReferences": [],
            "cDeriver": f"perf-{suffix}.drv",
            "cSig": "perfsig",
        }
    }
    res = await client.post(url, json=body, timeout=timeout)
    res.raise_for_status()
    return res


async def async_perform_download(client, file_hash, timeout=120):
    url = f"{BASE_URL}/nar/{file_hash}.nar.xz"
    res = await client.get(url, timeout=timeout)
    res.raise_for_status()
    return res


async def run_single_worker(index, base_data, client, loop):
    """Perform Upload -> Complete -> Download with random jitter to ensure mixing."""
    # 0. Random jitter (0 to 3 seconds) to desynchronize workers
    await asyncio.sleep(random.uniform(0.1, 0.5))

    current_step = "INIT"
    try:
        # 1. Initialize
        upload_id = await async_init_upload(client)

        # 2. Prepare unique data (Offload SHA256 to thread to avoid blocking event loop)
        # Hashing 100MB is CPU-intensive.
        data, file_hash = await loop.run_in_executor(
            None, get_unique_data, base_data, "para", index
        )
        actual_size = len(data)

        current_step = "UPLOAD"
        # 3. Upload (PUT)
        await async_perform_upload(client, upload_id, data)

        current_step = "COMPLEATE"
        # 4. Complete (Rename to hash)
        await async_complete_upload(
            client, upload_id, file_hash, actual_size, f"para{index}"
        )

        current_step = "DOWNLOAD"
        # 5. Immediate Download (GET by hash)
        await async_perform_download(client, file_hash)

        current_step = "DONE"
        return True
    except httpx.HTTPStatusError as e:
        print(
            f"[!] Worker {index:02d} FAILED at [{current_step}]: HTTP {e.response.status_code} - {e.response.text[:200]}"
        )
        return False
    except httpx.RequestError as e:
        print(f"[!] Worker {index:02d} FAILED at [{current_step}]: {e.request}")
        # TODO: retry go to step
        return False
    except Exception as e:
        print(f"[!] Worker {index:02d} FAILED at [{current_step}]: {e}")
        return False


async def main():
    print(
        f"[Phase] Starting Shuffled Async Test ({CONCURRENCY}x {FILE_SIZE_MB}MB Mixed Flow)..."
    )

    # Pre-generate base data
    base_data = os.urandom(FILE_SIZE_MB * 1024 * 1024)

    start_time = time.time()
    loop = asyncio.get_running_loop()

    # Connection limits
    limits = httpx.Limits(
        max_connections=CONCURRENCY, max_keepalive_connections=CONCURRENCY
    )
    headers = {"Authorization": f"Bearer {KEY}"}

    async with httpx.AsyncClient(
        verify=False, limits=limits, headers=headers, timeout=None
    ) as client:
        tasks = [
            run_single_worker(i, base_data, client, loop) for i in range(CONCURRENCY)
        ]
        results = await asyncio.gather(*tasks)

    end_time = time.time()
    successful = [r for r in results if r]
    print(f"\n[Phase] Shuffled Async Test Finished in {end_time - start_time:.5f}s.")
    print(f"Final Score: {len(successful)}/{CONCURRENCY} successful.")


if __name__ == "__main__":
    asyncio.run(main())

File: api-test/k6t/concurrent_heavy.js
import http from 'k6/http';
import { check, sleep } from 'k6';
import crypto from 'k6/crypto';

// --- Configuration ---
const MGMT_URL = 'https://localhost/api/v1';
const CACHE_URL = 'https://test.localhost';
const AUTH_TOKEN = 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoidGVzdCJ9.Pijr3YxjmBgUGuPKox3zwxro6gjfgdoul36XFHUH1Ro';;
const CACHE_NAME = 'test';
const FILE_SIZE_MB = 100;

export const options = {
  stages: [
    { duration: '1m', target: 20 },
    { duration: '3m', target: 20 },
    { duration: '1m', target: 0 },
  ],
  thresholds: {
    http_req_failed: ['rate<0.05'],
  },
};

const sharedData = crypto.randomBytes(FILE_SIZE_MB * 1024 * 1024);

export default function () {
  const index = __VU;
  const iteration = __ITER;

  const fileHash = crypto.sha256(sharedData, 'hex');
  const actualSize = sharedData.byteLength;

  const params = {
    headers: {
      'Authorization': AUTH_TOKEN,
      'Content-Type': 'application/json',
    },
    timeout: '300s',
  };

  // --- Step 1: INIT ---
  const initRes = http.post(`${MGMT_URL}/cache/${CACHE_NAME}/multipart-nar?compression=xz`, null, params);
  if (!check(initRes, { 'init success': (r) => r.status === 200 })) return;
  const uploadId = initRes.json().uploadId;

  // --- Step 2: UPLOAD ---
  const uploadParams = {
    headers: {
      'Authorization': AUTH_TOKEN,
      'Content-Type': 'application/octet-stream',
    },
    timeout: '300s',
  };

  // We pass the shared buffer directly.
  const uploadRes = http.put(`${CACHE_URL}/${uploadId}`, sharedData, uploadParams);
  if (!check(uploadRes, { 'upload success': (r) => r.status === 201 })) return;

  // --- Step 3: COMPLETE ---
  // We append index and iteration to suffix to ensure unique DB entries even if data hash is same
  const completeBody = JSON.stringify({
    narInfoCreate: {
      cFileHash: fileHash,
      cFileSize: actualSize,
      cStoreHash: fileHash.substring(0, 32),
      cStoreSuffix: `k6-vu${index}-it${iteration}`,
      cNarHash: `sha256:${fileHash}`,
      cNarSize: actualSize,
      cReferences: [],
      cDeriver: `k6-vu${index}.drv`,
      cSig: 'perfsig',
    },
  });
  const completeRes = http.post(`${MGMT_URL}/cache/${CACHE_NAME}/multipart-nar/${uploadId}/complete`, completeBody, params);
  if (!check(completeRes, { 'complete success': (r) => r.status === 200 })) return;

  // --- Step 4: DOWNLOAD ---
  const downloadRes = http.get(`${CACHE_URL}/nar/${fileHash}.nar.xz`, params);
  check(downloadRes, { 'download success': (r) => r.status === 200 });

  sleep(Math.random() * 2 + 1);
}
